sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala

```diff
@@ -128,6 +128,58 @@ case class IsNaN(child: Expression) extends UnaryExpression
 }
 
 /**
+ * An Expression accepts two parameters and returns null if both parameters are equal.
+ * If they are not equal, the first parameter value is returned.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(a,b) - Returns null if a equals to b, or a otherwise.")
+case class NullIf(left: Expression, right: Expression) extends BinaryExpression {
+  override def nullable: Boolean = true
+  override def dataType: DataType = left.dataType
+
+  override def eval(input: InternalRow): Any = {
+    val valueLeft = left.eval(input)
+    val valueRight = right.eval(input)
+    if (valueLeft.equals(valueRight)) {
+      null
+    } else {
+      valueLeft
+    }
+  }
+
+  override def genCode(ctx: CodegenContext, ev: ExprCode): String = {
+    val leftGen = left.gen(ctx)
+    val rightGen = right.gen(ctx)
+    dataType match {
```

Comment:
We have `CodegenContext.genEqual` to generate codes for equality check. You can see how `EqualTo` expression uses it.